src/abstractdataframe/join.jl

```diff
@@ -6,21 +6,34 @@
 similar_missing(dv::AbstractArray{T}, dims::Union{Int, Tuple{Vararg{Int}}}) where {T} =
     fill!(similar(dv, Union{T, Missing}, dims), missing)
 
+const OnType = Union{Symbol, NTuple{2,Symbol}, Pair{Symbol,Symbol}}
+
 # helper structure for DataFrames joining
 struct DataFrameJoiner{DF1<:AbstractDataFrame, DF2<:AbstractDataFrame}
     dfl::DF1
     dfr::DF2
     dfl_on::DF1
     dfr_on::DF2
-    on_cols::Vector{Symbol}
+    left_on::Vector{Symbol}
+    right_on::Vector{Symbol}
+
+    function DataFrameJoiner{DF1, DF2}(dfl::DF1, dfr::DF2,
+                                       on::Union{<:OnType, AbstractVector{<:OnType}}) where {DF1, DF2}
 
-    function DataFrameJoiner{DF1, DF2}(dfl::DF1, dfr::DF2, on::Union{Symbol,Vector{Symbol}}) where {DF1, DF2}
-        on_cols = isa(on, Symbol) ? [on] : on
-        new(dfl, dfr, dfl[on_cols], dfr[on_cols], on_cols)
+        on_cols = isa(on, Vector) ? on : [on]
+        if eltype(on_cols) == Symbol
+            left_on = on_cols
+            right_on = on_cols
+        else
+            left_on = [first(x) for x in on_cols]
+            right_on = [last(x) for x in on_cols]
+        end
+        new(dfl, dfr, dfl[left_on], dfr[right_on], left_on, right_on)
     end
 end
 
-DataFrameJoiner(dfl::DF1, dfr::DF2, on::Union{Symbol,Vector{Symbol}}) where {DF1<:AbstractDataFrame, DF2<:AbstractDataFrame} =
+DataFrameJoiner(dfl::DF1, dfr::DF2, on::Union{<:OnType, AbstractVector{<:OnType}}) where
+    {DF1<:AbstractDataFrame, DF2<:AbstractDataFrame} =
```

Comment:
I guess that's the indentation convention for this kind of situation? I'm not used to it yet.

Reply:
I'm not sure if there is a convention, but this only works if the line break is after the `where` clause. If you line break before the `where` then it results in an error

Reply:
Yeah, that's annoying. But I was wondering whether we should use four spaces, or align with arguments. Hard to decide.